### PR TITLE
koji_promote: store help markdown filename in the build metadata if available

### PR DESCRIPTION
**Pull request by vrutkovs**

### CODE REVIEW

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 1]
```diff
@@ -560,6 +561,17 @@ def get_build(self, metadata):
                 else:
                     extra['filesystem_koji_task_id'] = task_id
 
+        help_result = self.workflow.prebuild_results.get(AddHelpPlugin.key)
+        if help_result is not None:
+            try:
+                cgit_url = self.workflow.source.uri.replace('/rpms/', '/cgit/rpms/')
```
Comment (lcarva) [inline atomic_reactor/plugins/exit_koji_promote.py]: This seems too specific to our deployment.

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 2]
```diff
@@ -560,6 +561,17 @@ def get_build(self, metadata):
...
+                help_md = self.workflow.prebuild_plugins_conf[AddHelpPlugin.key]['help_file']
+                commit = self.workflow.source.lg.commit_id
+                extra['help'] = "{cgit_url}/tree/{help_md}?id={commit}".format(
```
Comment (lcarva) [inline atomic_reactor/plugins/exit_koji_promote.py]: This seems too specific to our deployment.

File: atomic_reactor/plugins/exit_koji_promote.py (hunk 1 above)
Comment (lcarva) [inline atomic_reactor/plugins/exit_koji_promote.py]: I think this can be solved by modifying this plugin, or `AddHelpPlugin??`, to take a new param that specifies the template to be used to construct the URL.

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: I don't know if we even want a cgit url at all. That's not what's discussed in the story.

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: This probably should be a part of `GitSource` class - it might try to detect the git frontend

Comment (lcarva) [inline atomic_reactor/plugins/exit_koji_promote.py]: A template string should allow users to use whatever hosting solution they'd like. We could format it with what we know: `commit`, `help_file_name`, others?

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: We already have "the git URL" -- I was thinking we'd use that. There is a list of required fields in the story description.

File: atomic_reactor/plugins/exit_store_metadata_in_osv3.py [hunk 3]
```diff
@@ -157,7 +157,7 @@ def run(self):
 
         annotations = {
             "dockerfile": open(self.workflow.builder.df_path).read(),
-            "artefacts": self.get_pre_result(DistgitFetchArtefactsPlugin.key),
+            "artefacts": self.get_pre_result(DistgitFetchArtefactsPlugin.key) or [],
```
Comment (lcarva) [inline atomic_reactor/plugins/exit_store_metadata_in_osv3.py]: How about remove this key altogether?

File: atomic_reactor/plugins/exit_koji_promote.py (hunk 1 above)
Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: How about formatting it as `{{git_url}}#{{commit}};{{filename}}`? Do we know if other koji plugins are using any scheme?

Sorry for discussing this after implementing, I expected the task to detect a git frontend to be much simplier

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: The requirement is that the metadata can be used to discover the dist-git URL, commit hash, and pathname to the Markdown source file.

It needn't be a single field.

Note that build.source already includes the dist-git URL and commit hash, so I think all that's needed is to include the path within the repository in the build.extra.image map.

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: >Note that build.source already includes the dist-git URL and commit hash, so I think all that's needed is to include the path within the repository in the build.extra.image map.

Agreed, implemented

**Comment by lcarva** [comment]
How about rewording the commit message and PR title?

**Comment by twaugh** [comment]
Once https://github.com/projectatomic/atomic-reactor/pull/622 is merged this could update docs/koji.md.

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 4]
```diff
@@ -560,6 +561,13 @@ def get_build(self, metadata):
                 else:
                     extra['filesystem_koji_task_id'] = task_id
 
+        help_result = self.workflow.prebuild_results.get(AddHelpPlugin.key)
+        if help_result is not None:
+            try:
+                extra['help'] = self.workflow.prebuild_plugins_conf[AddHelpPlugin.key]['help_file']
+            except:
+                extra['help'] = DEFAULT_HELP_FILENAME
```
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: Although it appears to be the correct place because we have existing keys there, I think technically it ought to live in extra['image']['help'].

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Fixed

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 5]
```diff
@@ -560,6 +561,13 @@ def get_build(self, metadata):
                 else:
                     extra['filesystem_koji_task_id'] = task_id
 
+        help_result = self.workflow.prebuild_results.get(AddHelpPlugin.key)
+        if help_result is not None:
+            try:
+                extra['image']['help'] = self.workflow.prebuild_plugins_conf[AddHelpPlugin.key]['help_file']
+            except:
+                extra['image']['help'] = DEFAULT_HELP_FILENAME
```
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: Could set this at line 539, like in https://github.com/projectatomic/atomic-reactor/pull/619?

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Yeah, it would be set there.

Do we want a key with an empty value if help file was not found? I guess not, right?

Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: Good question. Possible ways of handling this:

- empty string, so an empty string means there is no help to find
- use `None`, so that a null value means there is no help to find
- omit the key altogether, so the absence of the key means there is no help to find

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Updated the code to have `help` set to `None`

File: atomic_reactor/plugins/exit_koji_promote.py [hunk 6]
```diff
@@ -560,6 +561,13 @@ def get_build(self, metadata):
                 else:
                     extra['filesystem_koji_task_id'] = task_id
 
+        help_result = self.workflow.prebuild_results.get(AddHelpPlugin.key)
+        if help_result is not None:
+            try:
+                extra['image']['help'] = self.workflow.prebuild_plugins_conf[AddHelpPlugin.key]['help_file']
+            except:
```
Comment (twaugh) [inline atomic_reactor/plugins/exit_koji_promote.py]: `except KeyError:`?

Comment (vrutkovs) [inline atomic_reactor/plugins/exit_koji_promote.py]: Right, fixed

**Comment by twaugh** [comment]
Should the commit message say "build metadata" instead of "build config"?

File: docs/koji.md [hunk 7]
```diff
@@ -23,6 +23,7 @@ For atomic-reactor container image builds the `image` type is used, and so type-
 Data which is placed here includes:
 
 - `build.extra.image.autorebuild` (boolean): true if this build was triggered automatically; false otherwise
+- `build.extra.image.help` (boolean): filename if this build has a markdown help converted; None otherwise
```
Comment (twaugh) [inline docs/koji.md]: Says 'boolean' but should say 'string or null'. Also let's call it 'null' instead of 'None' as this is a JSON object.

Comment (vrutkovs) [inline docs/koji.md]: Updated

File: docs/koji.md [hunk 8]
```diff
@@ -23,6 +23,7 @@ For atomic-reactor container image builds the `image` type is used, and so type-
 Data which is placed here includes:
 
 - `build.extra.image.autorebuild` (boolean): true if this build was triggered automatically; false otherwise
+- `build.extra.image.help` (string or null): filename if this build has a markdown help converted; null otherwise
```
Comment (twaugh) [inline docs/koji.md]: I feel like we need to make it clear that it is the path to the Markdown file in the git repository (rather than to the converted file in the built image).

Comment (vrutkovs) [inline docs/koji.md]: Does "filename of the markdown help file in the repository if this build has a markdown help converted to man page; null otherwise" sound good?

Comment (twaugh) [inline docs/koji.md]: Perfect.

Comment (vrutkovs) [inline docs/koji.md]: Fixed

File: docs/koji.md [hunk 9]
```diff
@@ -61,6 +62,8 @@ Example type-specific content generator metadata in context:
     "release": "1",
     "extra": {
       "image": {
+        "autorebuild": "False",
+        "help": None,
```
Comment (twaugh) [inline docs/koji.md]: Ah, good catch! Yes, we need to update this. Probably should read like this though:
```
{
  "autorebuild": false,
  "help": null
},
```
(no comma after the final entry)

Comment (vrutkovs) [inline docs/koji.md]: Fixed

File: docs/koji.md [hunk 10]
```diff
@@ -61,6 +62,8 @@ Example type-specific content generator metadata in context:
     "release": "1",
     "extra": {
       "image": {
+        "autorebuild": "False",
```
Comment (twaugh) [inline docs/koji.md]: `false`

Comment (vrutkovs) [inline docs/koji.md]: Ah, weird, I though I updated
